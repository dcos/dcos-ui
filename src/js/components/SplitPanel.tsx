import * as React from "react";
import classNames from "classnames";
import { Icon } from "@dcos/ui-kit";
import { SystemIcons } from "@dcos/ui-kit/dist/packages/icons/dist/system-icons-enum";
import { iconSizeXxs } from "@dcos/ui-kit/dist/packages/design-tokens/build/js/designTokens";
import WindowResizeUtil from "#SRC/js/utils/WindowResizeUtil";

type SplitPanelChildren<T> = [T, T];

interface PanelProps {
  children: React.ReactNode;
  className?: string;
  isResizing?: boolean;
}

interface SidePanelProps extends PanelProps {
  initialWidth?: number;
  isActive?: boolean;
  width?: number;
}

export interface SplitPanelState {
  isResizing: boolean;
  sidePanelWidth?: number;
}

export interface SplitPanelProps {
  children: SplitPanelChildren<
    React.ReactElement<PanelProps> | React.ReactElement<SidePanelProps>
  >;
  onResize?: (resizedPanelWidth?: number) => void;
}

export const getSidePanelWidth = (
  containerRect: ClientRect,
  clientPosition: number
) => {
  const totalSize = containerRect.width;
  let offset;

  offset = clientPosition - containerRect.left;

  if (offset < 0) {
    offset = 0;
  } else if (offset > totalSize) {
    offset = totalSize;
  }

  return totalSize - offset;
};

export const getSidePanelProps = (
  panelChildren: SplitPanelChildren<
    React.ReactElement<PanelProps> | React.ReactElement<SidePanelProps>
  >
) => {
  const sidePanelChild = panelChildren.find(
    child => React.isValidElement(child) && child.type === SidePanel
  ) as React.ReactElement<SidePanelProps>;
  return sidePanelChild.props;
};

export const SidePanel = ({
  children,
  className,
  isActive,
  isResizing,
  width
}: SidePanelProps) => {
  const sidePanelClasses = classNames("modal-full-screen-side-panel", {
    "is-visible": isActive
  });
  return (
    <React.Fragment>
      <div
        className={classNames("modal-full-screen-side-panel-placeholder", {
          "is-visible": isActive
        })}
        style={{
          width: isActive ? width : undefined,
          transition: isResizing ? "none" : undefined
        }}
      />
      <div
        className={classNames(
          "splitPanels-panel",
          sidePanelClasses,
          className,
          {
            noUserSelect: isResizing
          }
        )}
        style={{
          width,
          right: isActive ? 0 : width && width * -1
        }}
      >
        {children}
      </div>
    </React.Fragment>
  );
};

export const PrimaryPanel = ({
  children,
  className,
  isResizing
}: PanelProps) => (
  <div
    className={classNames(
      "splitPanels-panel",
      "splitPanels-panel--primary",
      className,
      {
        noUserSelect: isResizing
      }
    )}
  >
    {children}
  </div>
);

function intersperse<A>(list: A[], sep: JSX.Element) {
  return Array.prototype.concat(...list.map(e => [sep, e])).slice(1);
}

class SplitPanel extends React.PureComponent<SplitPanelProps, SplitPanelState> {
  public containerRef = React.createRef<HTMLDivElement>();

  constructor(props: SplitPanelProps) {
    super(props);

    this.handleMouseDown = this.handleMouseDown.bind(this);
    this.handleMouseUp = this.handleMouseUp.bind(this);
    this.handleMouseMove = this.handleMouseMove.bind(this);
    this.handleWindowResize = this.handleWindowResize.bind(this);

    this.state = {
      isResizing: false
    };
  }

  public componentDidMount() {
    const { initialWidth } = getSidePanelProps(this.props.children);
    WindowResizeUtil.add(this.handleWindowResize);

    this.setState({ sidePanelWidth: initialWidth });
  }

  public componentWillUnmount() {
    WindowResizeUtil.remove(this.handleWindowResize);
  }

  public render() {
    const { children } = this.props;
    const { isResizing, sidePanelWidth } = this.state;
    const { isActive } = getSidePanelProps(this.props.children);
    const panelChildren = React.Children.toArray(children) as Array<
      React.ReactElement<PanelProps> & React.ReactElement<SidePanelProps>
    >;
    const panelChildrenArr = panelChildren.map(panel => {
      if (React.isValidElement(panel)) {
        if (panel.type === SidePanel) {
          return React.cloneElement(panel, {
            isResizing,
            width: sidePanelWidth,
            key: "sidePanel"
          });
        }
        if (panel.type === PrimaryPanel) {
          return React.cloneElement(panel, {
            isResizing,
            key: "primaryPanel"
          });
        }
      }
      return panel;
    });

    const panelSeparator = (
      <div
        className={classNames("splitPanels-separator", {
          "splitPanels-separator--active": isActive
        })}
        style={{
          transform: `translateX(${isActive ? 0 : sidePanelWidth}px)`,
          right: isActive ? sidePanelWidth : 0,
          transition: isResizing ? "none" : undefined
        }}
        onMouseDown={this.handleMouseDown}
        key="separator"
      >
        <div className="splitPanels-resizeIconWrapper">
          <Icon shape={SystemIcons.ResizeHorizontal} size={iconSizeXxs} />
        </div>
      </div>
    );

    return (
      <div
        className="splitPanels"
        onMouseMove={this.handleMouseMove}
        onMouseUp={this.handleMouseUp}
        ref={this.containerRef}
      >
        {intersperse(panelChildrenArr, panelSeparator)}
      </div>
    );
  }

  private handleMouseDown() {
    this.setState({ isResizing: true });
  }

  private handleMouseUp() {
    if (this.state.isResizing && this.props.onResize) {
      this.props.onResize(this.state.sidePanelWidth);
    }
    this.setState({ isResizing: false });
  }

  private handleMouseMove(e: React.MouseEvent<HTMLDivElement, MouseEvent>) {
    const containerRefCurrent = this.containerRef.current;
    if (this.state.isResizing && containerRefCurrent) {
      const sidePanelWidth = getSidePanelWidth(
        containerRefCurrent.getBoundingClientRect(),
        e.clientX
      );

      this.setState({ sidePanelWidth });
    }
  }

  private handleWindowResize() {
    const { body } = document;
    const { sidePanelWidth } = this.state;

    if (sidePanelWidth && body.clientWidth < sidePanelWidth) {
      this.setState({ sidePanelWidth: body.clientWidth });
    }
  }
}

export default SplitPanel;
