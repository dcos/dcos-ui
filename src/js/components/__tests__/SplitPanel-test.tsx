import * as React from "react";
import { shallow, mount } from "enzyme";

import SplitPanel, {
  PrimaryPanel,
  SidePanel,
  getSidePanelWidth,
  getSidePanelProps
} from "../SplitPanel";

describe("SplitPanel", () => {
  it("renders default", () => {
    const component = shallow(
      <SplitPanel>
        <PrimaryPanel>PrimaryPanel</PrimaryPanel>
        <SidePanel>SidePanel</SidePanel>
      </SplitPanel>
    );
    expect(component).toMatchSnapshot();
  });
  it("renders with SidePanel isActive prop set to true", () => {
    const component = shallow(
      <SplitPanel>
        <PrimaryPanel>PrimaryPanel</PrimaryPanel>
        <SidePanel isActive={true}>SidePanel</SidePanel>
      </SplitPanel>
    );
    expect(component).toMatchSnapshot();
  });
  it("renders SidePanel and PrimaryPanel with className passed", () => {
    const component = shallow(
      <SplitPanel>
        <PrimaryPanel className="className">PrimaryPanel</PrimaryPanel>
        <SidePanel className="className">SidePanel</SidePanel>
      </SplitPanel>
    );
    expect(component).toMatchSnapshot();
  });

  describe("resizing", () => {
    const mockContainerWidth = 100;
    const resizeAmount = 10;

    beforeEach(() => {
      Element.prototype.getBoundingClientRect = jest.fn(() => ({
        width: mockContainerWidth,
        height: 0,
        top: 0,
        left: 0,
        bottom: 0,
        right: 0
      }));
    });
    it("calls onResize callback when finished dragging", () => {
      const onResizeCallback = jest.fn();
      const component = mount(
        <SplitPanel onResize={onResizeCallback}>
          <PrimaryPanel>PrimaryPanel</PrimaryPanel>
          <SidePanel isActive={true}>SidePanel</SidePanel>
        </SplitPanel>
      );
      const separator = component.find(".splitPanels-separator");
      const panelContainer = component.find(".splitPanels"); // target the documentElement

      Element.prototype.getBoundingClientRect = jest.fn(() => ({
        width: mockContainerWidth,
        height: 0,
        top: 0,
        left: 0,
        bottom: 0,
        right: 0
      }));

      separator.simulate("mousedown");
      panelContainer.simulate("mousemove", { clientX: resizeAmount });
      separator.simulate("mouseup");
      expect(onResizeCallback).toHaveBeenCalledWith(
        mockContainerWidth - resizeAmount
      );
    });
    describe("#getSidePanelWidth", () => {
      it("returns difference between drag delta and full component width in pixels", () => {
        const element = document.createElement("span");
        const rect = element.getBoundingClientRect();
        expect(getSidePanelWidth(rect, resizeAmount)).toBe(
          mockContainerWidth - resizeAmount
        );
      });
      it("returns 0 if difference between drag delta and full component width is =< 0", () => {
        const element = document.createElement("span");
        const rect = element.getBoundingClientRect();
        expect(getSidePanelWidth(rect, mockContainerWidth + resizeAmount)).toBe(
          0
        );
      });
      it("returns full component width if difference between drag delta is >= full component width", () => {
        const element = document.createElement("span");
        const rect = element.getBoundingClientRect();
        expect(
          getSidePanelWidth(
            rect,
            parseInt((mockContainerWidth + resizeAmount) * -1, 10)
          )
        ).toBe(100);
      });
    });
  });
  describe("#getSidePanelProps", () => {
    it("returns the props passed to the SidePanel child", () => {
      const component = shallow(
        <SplitPanel>
          <PrimaryPanel>PrimaryPanel</PrimaryPanel>
          <SidePanel isActive={true}>SidePanel</SidePanel>
        </SplitPanel>
      );
      expect(component.find(SidePanel).props()).toEqual(
        getSidePanelProps(component.prop("children"))
      );
    });
  });
});
