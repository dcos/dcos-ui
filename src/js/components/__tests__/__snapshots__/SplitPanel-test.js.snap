// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SplitPanel renders SidePanel and PrimaryPanel with className passed 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SplitPanel>
    <Unknown
      className="className"
    >
      PrimaryPanel
    </Unknown>
    <Unknown
      className="className"
    >
      SidePanel
    </Unknown>
  </SplitPanel>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Unknown
          className="className"
          isResizing={false}
        >
          PrimaryPanel
        </Unknown>,
        <div
          className="splitPanels-separator"
          onMouseDown={[Function]}
          style={
            Object {
              "right": 0,
              "transform": "translateX(undefinedpx)",
              "transition": undefined,
            }
          }
        >
          <div
            className="splitPanels-resizeIconWrapper"
          >
            <Icon
              shape="system-resize-horizontal"
              size="12px"
            />
          </div>
        </div>,
        <Unknown
          className="className"
          isResizing={false}
          width={undefined}
        >
          SidePanel
        </Unknown>,
      ],
      "className": "splitPanels",
      "onMouseMove": [Function],
      "onMouseUp": [Function],
    },
    "ref": Object {
      "current": null,
    },
    "rendered": Array [
      Object {
        "instance": null,
        "key": "primaryPanel",
        "nodeType": "function",
        "props": Object {
          "children": "PrimaryPanel",
          "className": "className",
          "isResizing": false,
        },
        "ref": null,
        "rendered": "PrimaryPanel",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "separator",
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="splitPanels-resizeIconWrapper"
          >
            <Icon
              shape="system-resize-horizontal"
              size="12px"
            />
          </div>,
          "className": "splitPanels-separator",
          "onMouseDown": [Function],
          "style": Object {
            "right": 0,
            "transform": "translateX(undefinedpx)",
            "transition": undefined,
          },
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <Icon
              shape="system-resize-horizontal"
              size="12px"
            />,
            "className": "splitPanels-resizeIconWrapper",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "shape": "system-resize-horizontal",
              "size": "12px",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "div",
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": "sidePanel",
        "nodeType": "function",
        "props": Object {
          "children": "SidePanel",
          "className": "className",
          "isResizing": false,
          "width": undefined,
        },
        "ref": null,
        "rendered": "SidePanel",
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Unknown
            className="className"
            isResizing={false}
          >
            PrimaryPanel
          </Unknown>,
          <div
            className="splitPanels-separator"
            onMouseDown={[Function]}
            style={
              Object {
                "right": 0,
                "transform": "translateX(undefinedpx)",
                "transition": undefined,
              }
            }
          >
            <div
              className="splitPanels-resizeIconWrapper"
            >
              <Icon
                shape="system-resize-horizontal"
                size="12px"
              />
            </div>
          </div>,
          <Unknown
            className="className"
            isResizing={false}
            width={undefined}
          >
            SidePanel
          </Unknown>,
        ],
        "className": "splitPanels",
        "onMouseMove": [Function],
        "onMouseUp": [Function],
      },
      "ref": Object {
        "current": null,
      },
      "rendered": Array [
        Object {
          "instance": null,
          "key": "primaryPanel",
          "nodeType": "function",
          "props": Object {
            "children": "PrimaryPanel",
            "className": "className",
            "isResizing": false,
          },
          "ref": null,
          "rendered": "PrimaryPanel",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "separator",
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="splitPanels-resizeIconWrapper"
            >
              <Icon
                shape="system-resize-horizontal"
                size="12px"
              />
            </div>,
            "className": "splitPanels-separator",
            "onMouseDown": [Function],
            "style": Object {
              "right": 0,
              "transform": "translateX(undefinedpx)",
              "transition": undefined,
            },
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <Icon
                shape="system-resize-horizontal"
                size="12px"
              />,
              "className": "splitPanels-resizeIconWrapper",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "shape": "system-resize-horizontal",
                "size": "12px",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": "div",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": "sidePanel",
          "nodeType": "function",
          "props": Object {
            "children": "SidePanel",
            "className": "className",
            "isResizing": false,
            "width": undefined,
          },
          "ref": null,
          "rendered": "SidePanel",
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;

exports[`SplitPanel renders default 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SplitPanel>
    <Unknown>
      PrimaryPanel
    </Unknown>
    <Unknown>
      SidePanel
    </Unknown>
  </SplitPanel>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Unknown
          isResizing={false}
        >
          PrimaryPanel
        </Unknown>,
        <div
          className="splitPanels-separator"
          onMouseDown={[Function]}
          style={
            Object {
              "right": 0,
              "transform": "translateX(undefinedpx)",
              "transition": undefined,
            }
          }
        >
          <div
            className="splitPanels-resizeIconWrapper"
          >
            <Icon
              shape="system-resize-horizontal"
              size="12px"
            />
          </div>
        </div>,
        <Unknown
          isResizing={false}
          width={undefined}
        >
          SidePanel
        </Unknown>,
      ],
      "className": "splitPanels",
      "onMouseMove": [Function],
      "onMouseUp": [Function],
    },
    "ref": Object {
      "current": null,
    },
    "rendered": Array [
      Object {
        "instance": null,
        "key": "primaryPanel",
        "nodeType": "function",
        "props": Object {
          "children": "PrimaryPanel",
          "isResizing": false,
        },
        "ref": null,
        "rendered": "PrimaryPanel",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "separator",
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="splitPanels-resizeIconWrapper"
          >
            <Icon
              shape="system-resize-horizontal"
              size="12px"
            />
          </div>,
          "className": "splitPanels-separator",
          "onMouseDown": [Function],
          "style": Object {
            "right": 0,
            "transform": "translateX(undefinedpx)",
            "transition": undefined,
          },
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <Icon
              shape="system-resize-horizontal"
              size="12px"
            />,
            "className": "splitPanels-resizeIconWrapper",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "shape": "system-resize-horizontal",
              "size": "12px",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "div",
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": "sidePanel",
        "nodeType": "function",
        "props": Object {
          "children": "SidePanel",
          "isResizing": false,
          "width": undefined,
        },
        "ref": null,
        "rendered": "SidePanel",
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Unknown
            isResizing={false}
          >
            PrimaryPanel
          </Unknown>,
          <div
            className="splitPanels-separator"
            onMouseDown={[Function]}
            style={
              Object {
                "right": 0,
                "transform": "translateX(undefinedpx)",
                "transition": undefined,
              }
            }
          >
            <div
              className="splitPanels-resizeIconWrapper"
            >
              <Icon
                shape="system-resize-horizontal"
                size="12px"
              />
            </div>
          </div>,
          <Unknown
            isResizing={false}
            width={undefined}
          >
            SidePanel
          </Unknown>,
        ],
        "className": "splitPanels",
        "onMouseMove": [Function],
        "onMouseUp": [Function],
      },
      "ref": Object {
        "current": null,
      },
      "rendered": Array [
        Object {
          "instance": null,
          "key": "primaryPanel",
          "nodeType": "function",
          "props": Object {
            "children": "PrimaryPanel",
            "isResizing": false,
          },
          "ref": null,
          "rendered": "PrimaryPanel",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "separator",
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="splitPanels-resizeIconWrapper"
            >
              <Icon
                shape="system-resize-horizontal"
                size="12px"
              />
            </div>,
            "className": "splitPanels-separator",
            "onMouseDown": [Function],
            "style": Object {
              "right": 0,
              "transform": "translateX(undefinedpx)",
              "transition": undefined,
            },
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <Icon
                shape="system-resize-horizontal"
                size="12px"
              />,
              "className": "splitPanels-resizeIconWrapper",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "shape": "system-resize-horizontal",
                "size": "12px",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": "div",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": "sidePanel",
          "nodeType": "function",
          "props": Object {
            "children": "SidePanel",
            "isResizing": false,
            "width": undefined,
          },
          "ref": null,
          "rendered": "SidePanel",
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;

exports[`SplitPanel renders with SidePanel isActive prop set to true 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SplitPanel>
    <Unknown>
      PrimaryPanel
    </Unknown>
    <Unknown
      isActive={true}
    >
      SidePanel
    </Unknown>
  </SplitPanel>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Unknown
          isResizing={false}
        >
          PrimaryPanel
        </Unknown>,
        <div
          className="splitPanels-separator splitPanels-separator--active"
          onMouseDown={[Function]}
          style={
            Object {
              "right": undefined,
              "transform": "translateX(0px)",
              "transition": undefined,
            }
          }
        >
          <div
            className="splitPanels-resizeIconWrapper"
          >
            <Icon
              shape="system-resize-horizontal"
              size="12px"
            />
          </div>
        </div>,
        <Unknown
          isActive={true}
          isResizing={false}
          width={undefined}
        >
          SidePanel
        </Unknown>,
      ],
      "className": "splitPanels",
      "onMouseMove": [Function],
      "onMouseUp": [Function],
    },
    "ref": Object {
      "current": null,
    },
    "rendered": Array [
      Object {
        "instance": null,
        "key": "primaryPanel",
        "nodeType": "function",
        "props": Object {
          "children": "PrimaryPanel",
          "isResizing": false,
        },
        "ref": null,
        "rendered": "PrimaryPanel",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "separator",
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="splitPanels-resizeIconWrapper"
          >
            <Icon
              shape="system-resize-horizontal"
              size="12px"
            />
          </div>,
          "className": "splitPanels-separator splitPanels-separator--active",
          "onMouseDown": [Function],
          "style": Object {
            "right": undefined,
            "transform": "translateX(0px)",
            "transition": undefined,
          },
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <Icon
              shape="system-resize-horizontal"
              size="12px"
            />,
            "className": "splitPanels-resizeIconWrapper",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "shape": "system-resize-horizontal",
              "size": "12px",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "div",
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": "sidePanel",
        "nodeType": "function",
        "props": Object {
          "children": "SidePanel",
          "isActive": true,
          "isResizing": false,
          "width": undefined,
        },
        "ref": null,
        "rendered": "SidePanel",
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Unknown
            isResizing={false}
          >
            PrimaryPanel
          </Unknown>,
          <div
            className="splitPanels-separator splitPanels-separator--active"
            onMouseDown={[Function]}
            style={
              Object {
                "right": undefined,
                "transform": "translateX(0px)",
                "transition": undefined,
              }
            }
          >
            <div
              className="splitPanels-resizeIconWrapper"
            >
              <Icon
                shape="system-resize-horizontal"
                size="12px"
              />
            </div>
          </div>,
          <Unknown
            isActive={true}
            isResizing={false}
            width={undefined}
          >
            SidePanel
          </Unknown>,
        ],
        "className": "splitPanels",
        "onMouseMove": [Function],
        "onMouseUp": [Function],
      },
      "ref": Object {
        "current": null,
      },
      "rendered": Array [
        Object {
          "instance": null,
          "key": "primaryPanel",
          "nodeType": "function",
          "props": Object {
            "children": "PrimaryPanel",
            "isResizing": false,
          },
          "ref": null,
          "rendered": "PrimaryPanel",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "separator",
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="splitPanels-resizeIconWrapper"
            >
              <Icon
                shape="system-resize-horizontal"
                size="12px"
              />
            </div>,
            "className": "splitPanels-separator splitPanels-separator--active",
            "onMouseDown": [Function],
            "style": Object {
              "right": undefined,
              "transform": "translateX(0px)",
              "transition": undefined,
            },
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <Icon
                shape="system-resize-horizontal"
                size="12px"
              />,
              "className": "splitPanels-resizeIconWrapper",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "shape": "system-resize-horizontal",
                "size": "12px",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": "div",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": "sidePanel",
          "nodeType": "function",
          "props": Object {
            "children": "SidePanel",
            "isActive": true,
            "isResizing": false,
            "width": undefined,
          },
          "ref": null,
          "rendered": "SidePanel",
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
